TestInvocationShouldNotHang_whenClientShutsDown

```
11:05:59 --- FAIL: TestInvocationShouldNotHang_whenClientShutsDown (60.06s)
11:05:59 	assertions.go:29: true != false : invocationNotSent failed
```


This test sometimes fails.

--- a/internal/connection_manager.go
+++ b/internal/connection_manager.go
@@ -106,6 +106,9 @@ func (cm *connectionManagerImpl) onConnectionClose(connection *Connection, cause
 }
 
 func (cm *connectionManagerImpl) getOrTriggerConnect(address core.Address) (*Connection, error) {
+	if cm.isAlive.Load() == false {
+		return nil, core.NewHazelcastClientNotActiveError("Connection Manager is not active", nil)
+	}
 	connection := cm.getConnection(address, false)
 	if connection != nil {
 		return connection, nil
@@ -131,8 +134,13 @@ func (cm *connectionManagerImpl) getOwnerConnection() *Connection {
 }
 
 func (cm *connectionManagerImpl) shutdown() {
+	if cm.isAlive.Load() == false {
+		return
+	}
+	cm.isAlive.Store(false)
 	activeCons := cm.getActiveConnections()
 	for _, con := range activeCons {
+
 		con.close(core.NewHazelcastClientNotActiveError("client is shutting down", nil))
 	}
 }
@@ -147,6 +155,7 @@ type connectionManagerImpl struct {
 	listenerMutex       sync.Mutex
 	connectionListeners atomic.Value
 	addressTranslator   AddressTranslator
+	isAlive             atomic.Value
 }
 
 func newConnectionManager(client *HazelcastClient, addressTranslator AddressTranslator) connectionManager {
@@ -156,6 +165,7 @@ func newConnectionManager(client *HazelcastClient, addressTranslator AddressTran
 		addressTranslator: addressTranslator,
 	}
 	cm.connectionListeners.Store(make([]connectionListener, 0))
+	cm.isAlive.Store(true)
 	return &cm
 }
 
@@ -279,6 +289,10 @@ func (cm *connectionManagerImpl) createConnection(address core.Address, asOwner 
 	if !asOwner && cm.client.ClusterService.getOwnerConnectionAddress() == nil {
 		return nil, core.NewHazelcastIllegalStateError("ownerConnection is not active", nil)
 	}
+	if cm.isAlive.Load() == false {
+		return nil, core.NewHazelcastClientNotActiveError("Connection Manager is not active", nil)
+	}
+
 	invocationService := cm.client.InvocationService.(*invocationServiceImpl)
 	connectionID := cm.NextConnectionID()
 	con := newConnection(address, invocationService.handleResponse, connectionID, cm)

--- a/internal/invocation.go
+++ b/internal/invocation.go
@@ -74,6 +74,7 @@ func (i *invocation) complete(response interface{}) {
 	if atomic.CompareAndSwapInt32(&i.isComplete, 0, 1) {
 		i.response <- response
 	}
+
 }
 
 func (i *invocation) unwrapResponse(response interface{}) (*proto.ClientMessage, error) {
@@ -184,6 +185,7 @@ func (is *invocationServiceImpl) shutdown() {
 		delete(is.invocations, correlationID)
 		invocation.complete(core.NewHazelcastClientNotActiveError("client is shutting down", nil))
 	}
+
 	is.invocationsLock.Unlock()
 }
 

--- a/test/invocation_test.go
+++ b/test/invocation_test.go
@@ -150,7 +150,6 @@ func TestInvocationNotSent(t *testing.T) {
 					t.Fatal("put should have been retried, the error was :", err)
 				}
 			}()
-
 		}
 	}()
 	remoteController.ShutdownMember(cluster.ID, member.UUID)
@@ -183,7 +182,7 @@ func TestInvocationShouldNotHang_whenClientShutsDown(t *testing.T) {
 
 		}
 	}()
-	time.Sleep(10 * time.Millisecond)
+	time.Sleep(5 * time.Millisecond)
 	client.Shutdown()
 	timeout := WaitTimeout(wg, Timeout)
 	assert.Equalf(t, nil, timeout, false, "invocationNotSent failed")

